diff -aur a/src/glue-clipboard-macos.c b/src/glue-clipboard-macos.c
--- a/src/glue-clipboard-macos.c  2019-04-01 19:04:28.000000000 -0700
+++ b/src/glue-clipboard-macos.c  2019-04-03 15:12:20.000000000 -0700
@@ -29,6 +29,7 @@
 #include "glue-spice-widget-priv.h"
 #include "glue-clipboard.h"
 
+extern GMainContext *g_spice_glue_context;
 gboolean enableClipboardToGuest = FALSE;
 gboolean enableClipboardToClient = FALSE;
 uint32_t *guestClipboard = NULL;
@@ -122,7 +123,10 @@
         return 0;
     }
 
-    g_idle_add(grab_guest_clipboard, NULL);
+    GSource *source = g_idle_source_new();
+    g_source_set_callback(source, grab_guest_clipboard, NULL, NULL);
+    g_source_attach(source, g_spice_glue_context);
+    g_source_unref(source);
 
     return 0;
 }
@@ -156,7 +160,10 @@
         return 0;
     }
 
-    g_idle_add(release_guest_clipboard, NULL);
+    GSource *source = g_idle_source_new();
+    g_source_set_callback(source, release_guest_clipboard, NULL, NULL);
+    g_source_attach(source, g_spice_glue_context);
+    g_source_unref(source);
 
     return 0;
 }
@@ -192,7 +199,10 @@
         return 0;
     }
 
-    g_idle_add(clipboard_get_data, NULL);
+    GSource *source = g_idle_source_new();
+    g_source_set_callback(source, clipboard_get_data, NULL, NULL);
+    g_source_attach(source, g_spice_glue_context);
+    g_source_unref(source);
     
     return 1;
 }
diff -aur a/src/glue-clipboard-win32.c b/src/glue-clipboard-win32.c
--- a/src/glue-clipboard-win32.c  2019-04-01 19:04:28.000000000 -0700
+++ b/src/glue-clipboard-win32.c  2019-04-03 15:16:18.000000000 -0700
@@ -81,6 +81,8 @@
 GMutex data_mutex;
 GCond data_cond;
 
+extern GMainContext *g_spice_glue_context;
+
 //If more clipboard formats are to be supported, use CBData
 typedef struct
 {
@@ -503,9 +505,10 @@
     }
 
     if (wparam == CF_UNICODETEXT) {
-        g_timeout_add_full(G_PRIORITY_HIGH, 0,
-               onRenderFormat_text,
-               NULL, NULL);
+        GSource *source = g_timeout_source_new(0);
+        g_source_set_callback(source, onRenderFormat_text, NULL, NULL);
+        id = g_source_attach(source, g_spice_glue_context);
+        g_source_unref (source);
 /* https://msdn.microsoft.com/en-us/library/windows/desktop/ms649016(v=vs.85).aspx */
 
         // Allocate a buffer for the text. 
diff -aur a/src/glue-connection.c b/src/glue-connection.c
--- a/src/glue-connection.c 2019-04-01 19:04:28.000000000 -0700
+++ b/src/glue-connection.c 2019-04-02 13:13:31.000000000 -0700
@@ -245,6 +245,6 @@
 
 void spice_connection_power_event_request(SpiceConnection *conn, int powerEvent)
 {
-    if (conn->main != NULL)
-        spice_main_power_event_request(conn->main, powerEvent);
-}
\ No newline at end of file
+    //if (conn->main != NULL)
+    //    spice_main_power_event_request(conn->main, powerEvent);
+}
diff -aur a/src/glue-service.c b/src/glue-service.c
--- a/src/glue-service.c  2019-04-01 19:04:28.000000000 -0700
+++ b/src/glue-service.c  2019-04-03 13:33:21.000000000 -0700
@@ -72,6 +72,8 @@
 #include <stdlib.h>
 static FILE *logfile = NULL;
 
+GMainContext *g_spice_glue_context;
+
 void logToFile (const gchar *log_domain, GLogLevelFlags log_level,
    const gchar *message, gpointer user_data)
 {
@@ -179,7 +181,7 @@
 
 void SpiceGlibGlue_MainLoop(void)
 {
-    GMainLoop *mainloop = g_main_loop_new(NULL, FALSE);
+    GMainLoop *mainloop = g_main_loop_new(g_spice_glue_context, FALSE);
     g_main_loop_run(mainloop);
     g_main_loop_unref(mainloop);
 #if defined(PRINTING) || defined(SSO)
@@ -197,9 +199,10 @@
 
 void SpiceGlibGlue_Disconnect(void)
 {
-    g_timeout_add_full(G_PRIORITY_HIGH, 0,
-                       disconnect1,
-                       NULL, NULL);
+    GSource *source = g_timeout_source_new(0);
+    g_source_set_callback(source, disconnect1, NULL, NULL);
+    g_source_attach(source, g_spice_glue_context);
+    g_source_unref(source);
 }
 
 int16_t SpiceGlibGlue_Connect(char* host,
@@ -250,6 +253,7 @@
 
 void SpiceGlibGlueInitializeGlue()
 {
+    g_spice_glue_context = g_main_context_new();
 #ifdef PRINTING
     initializeFollowMePrinting();
 #endif
@@ -326,7 +330,8 @@
  *     in spice-protocol enums.h
  **/
 void SpiceGlibGlue_SendPowerEvent(int16_t powerEvent) {
-    g_timeout_add_full(G_PRIORITY_HIGH, 0,
-                       (GSourceFunc)sendPowerEvent1,
-                       (gpointer)powerEvent, NULL);
+    GSource *source = g_timeout_source_new(0);
+    g_source_set_callback(source, (GSourceFunc)sendPowerEvent1, (gpointer)powerEvent, NULL);
+    g_source_attach(source, g_spice_glue_context);
+    g_source_unref(source);
 }
diff -aur a/src/glue-spice-widget.c b/src/glue-spice-widget.c
--- a/src/glue-spice-widget.c 2019-04-01 19:04:28.000000000 -0700
+++ b/src/glue-spice-widget.c 2019-04-03 13:16:28.000000000 -0700
@@ -41,6 +41,7 @@
 #include "mono-glue-types.h"
 #include "glue-clipboard.h"
 
+extern GMainContext *g_spice_glue_context;
 
 static struct {
     uint32_t *buffer;
@@ -1063,7 +1064,10 @@
     }
 
     if (!d->copy_scheduled) {
-        g_timeout_add(30, (GSourceFunc) copy_display_to_glue, NULL);
+        GSource *source = g_timeout_source_new(30);
+        g_source_set_callback(source, (GSourceFunc) copy_display_to_glue, NULL, NULL);
+        g_source_attach(source, g_spice_glue_context);
+        g_source_unref(source);
         d->copy_scheduled = 1;
     }
 }
@@ -1623,7 +1627,10 @@
     if (global_display() != NULL) {
         SpiceDisplayPrivate *d = SPICE_DISPLAY_GET_PRIVATE(global_display());
         if (!d->copy_scheduled) {
-            g_timeout_add(30, (GSourceFunc) copy_display_to_glue, NULL);
+            GSource *source = g_timeout_source_new(30);
+            g_source_set_callback(source, (GSourceFunc) copy_display_to_glue, NULL, NULL);
+            g_source_attach(source, g_spice_glue_context);
+            g_source_unref(source);
             d->copy_scheduled = 1;
         }
     }
diff -aur a/src/usb-device-widget.c b/src/usb-device-widget.c
--- a/src/usb-device-widget.c 2019-04-01 19:04:28.000000000 -0700
+++ b/src/usb-device-widget.c 2019-04-03 15:16:12.000000000 -0700
@@ -25,6 +25,8 @@
 #include "usb-device-widget.h"
 #include <spice-gtk/spice-util-priv.h>
 
+extern GMainContext *g_spice_glue_context;
+
 /**
  * SECTION:usb-device-widget
  * @short_description: USB device selection glue
@@ -510,9 +512,10 @@
     connect_cb_data *data = g_new(connect_cb_data, 1);
     data->self  = g_object_ref(self);
     data->device  = device;
-    g_timeout_add_full(G_PRIORITY_HIGH, 0,
-                       spice_usb_device_widget_unshare1,
-                       data, NULL);
+    GSource *source = g_timeout_source_new(0);
+    g_source_set_callback(source, spice_usb_device_widget_unshare1, NULL, NULL);
+    id = g_source_attach(source, g_spice_glue_context);
+    g_source_unref (source);
 
 }
 
diff -aur a/src/usb-glue.c b/src/usb-glue.c
--- a/src/usb-glue.c  2019-04-01 19:04:28.000000000 -0700
+++ b/src/usb-glue.c  2019-04-03 15:16:05.000000000 -0700
@@ -108,9 +108,10 @@
 
 void SpiceGlibGlue_InitWindowsEvents() {
 
-    g_timeout_add_full(G_PRIORITY_HIGH, 0,
-                       startMessageLoop,
-                       NULL, NULL);
+    GSource *source = g_timeout_source_new(0);
+    g_source_set_callback(source, startMessageLoop, NULL, NULL);
+    id = g_source_attach(source, g_spice_glue_context);
+    g_source_unref (source);
 }
 
 static gboolean finalizeMessageWindow() {
@@ -122,9 +123,10 @@
 }
 
 void SpiceGlibGlue_FinalizeWindowsEvents() {
-    g_timeout_add_full(G_PRIORITY_HIGH, 0,
-                       finalizeMessageWindow,
-                       NULL, NULL);
+    GSource *source = g_timeout_source_new(0);
+    g_source_set_callback(source, finalizeMessageWindow, NULL, NULL);
+    id = g_source_attach(source, g_spice_glue_context);
+    g_source_unref (source);
 }
 
 #endif
